### PR TITLE
Sigopt with the new experiment framework

### PR DESCRIPTION
**Pull request by mvacaporale**
This reorganizes our sigopt utilities to make sure we can tune hyper-parameters easily and generally in the new Vernon framework
* Renames `SigoptImagenetTrainable` to `SigoptSupervisedTrainable`
* Introduces mixins for the SigoptExperiment class - for example `LRStepParams` is a mixin for tuning lr-step scheduler params
* There are mixins for lr-step and one-cycle schedulers and SGD optimizer params
* There is an `ImagenetParams` mixin for reproducibility of previous experiments
* Common experiments include `SigOptSDGStepLRExperiment` and `SigOptSDGOneCycleLRExperiment`
* A test config in the GSC project directory offers an example and exercises the new functionality. 

@subutai Could you please take a look at `nupic.research/nupic/research/frameworks/sigopt/mixins/imagenet_params.py`. This copies the original logic verbatim you used to run previous hyper-parameters searches. Just want to make sure this looks reasonable.

### CODE REVIEW

File: nupic/research/frameworks/sigopt/sigopt_experiment.py [hunk 1]
```diff
@@ -198,55 +178,11 @@ def update_config_with_suggestion(self, config, suggestion):
...
-            assignments.pop("init_lr", None)
-            assignments.pop("max_lr", None)
-
-        config.update(assignments)
+
+    @classmethod
+    def get_execution_order(cls):
+        return dict(
+            update_config_with_suggestion=[
+                "SupervisedExperiment.update_config_with_suggestion"
```
Comment (lscheinkman) [inline nupic/research/frameworks/sigopt/sigopt_experiment.py:186]: Did you mean "SigOptExperiment.update_config_with_suggestion"?

File: nupic/research/frameworks/sigopt/mixins/imagenet_params.py [hunk 2]
```diff
@@ -0,0 +1,131 @@
...
+            assignments.pop("init_lr", None)
+            assignments.pop("max_lr", None)
+
+        config.update(assignments)
+
+    @classmethod
+    def get_execution_order(cls):
+        eo = super().get_execution_order()
+        name = "ImagenetParams"
+        eo["update_config_with_suggestion"] = [name + ".update_config_with_suggestion"]
```
Comment (lscheinkman) [inline nupic/research/frameworks/sigopt/mixins/imagenet_params.py:130]: Why not `append`?

File: nupic/research/frameworks/sigopt/mixins/imagenet_params.py [hunk 3]
```diff
@@ -0,0 +1,131 @@
...
+#
+# http://numenta.org/licenses/
+# ----------------------------------------------------------------------
+
+import math
+
+
+class ImagenetParams(object):
+    """
+    This class is for backwards compatibility for older configs that train using
```
Comment (lscheinkman) [inline nupic/research/frameworks/sigopt/mixins/imagenet_params.py:27]: What do you mean by "older configs"? How would this change affect "newer configs"?

File: nupic/research/frameworks/sigopt/mixins/imagenet_params.py [hunk 4]
```diff
@@ -0,0 +1,131 @@
...
+        Given a SigOpt suggestion, update the optimizer and lr scheduler args.
+        Here, SGD is supported along with either StepLR or OneCycleLR.
+
+        :param config:
+            - lr_scheduler_args: dict of lr arguments
+            - lr_scheduler_class: OneCycleLR
+        :param suggestion:
+            - assignments (all optional)
+
+                # SDG Params
```
Comment (subutai) [inline nupic/research/frameworks/sigopt/mixins/imagenet_params.py:42]: "SGD"

File: nupic/research/frameworks/sigopt/mixins/imagenet_params.py [hunk 5]
```diff
@@ -0,0 +1,131 @@
...
+                - log_lr
+                - momentum
+                - weight_decay
+
+                # OneCylceLR Params
```
Comment (subutai) [inline nupic/research/frameworks/sigopt/mixins/imagenet_params.py:47]: "OneCycleLR"

File: nupic/research/frameworks/sigopt/mixins/one_cycle_lr_params.py [hunk 6]
```diff
@@ -0,0 +1,99 @@
...
+            max_epochs = self.sigopt_config["metadata"]["max_epochs"]
+            epochs = int(max_epochs * suggestion.task.cost)
+            lr_scheduler_args["epochs"] = epochs
+
+        if "momentum" in assignments:
+            lr_scheduler_args["max_momentum"] = assignments["momentum"]
+
+        if "pct_start" in assignments:
+            # Ensure integer number of epochs in pct start to avoid crash.
+            start_epochs = int(assignments["pct_start"] * config["epochs"] + 0.5)
```
Comment (subutai) [inline nupic/research/frameworks/sigopt/mixins/one_cycle_lr_params.py:68]: Wouldn't config["epochs"] be incorrect in a multi-task experiment?

File: nupic/research/frameworks/sigopt/mixins/one_cycle_lr_params.py [hunk 7]
```diff
@@ -0,0 +1,99 @@
...
+        lr_scheduler_class = config["lr_scheduler_class"]
+
+        assert lr_scheduler_class == OneCycleLR
+
+        # For multi-task experiments where epoch is the task. Must have a metadata
+        # field called max_epochs.
+        if suggestion.task is not None and "epoch" in suggestion.task.name:
+            max_epochs = self.sigopt_config["metadata"]["max_epochs"]
+            epochs = int(max_epochs * suggestion.task.cost)
+            lr_scheduler_args["epochs"] = epochs
```
Comment (subutai) [inline nupic/research/frameworks/sigopt/mixins/one_cycle_lr_params.py:61]: Shouldn't we also have:
``
            config["epochs"] = epochs
``
I didn't see that anywhere else.

File: nupic/research/frameworks/sigopt/mixins/sgd_params.py [hunk 8]
```diff
@@ -0,0 +1,70 @@
...
+import math
+
+from torch.optim import SGD
+
+
+class SGDParams(object):
+
+    def update_config_with_suggestion(self, config, suggestion):
+        """
+        Given a SigOpt suggestion, update the optimizer_args with SGD optimizer params.
```
Comment (subutai) [inline nupic/research/frameworks/sigopt/mixins/sgd_params.py:31]: Nothing in this class seems specific to SGD. Other optimizers could have learning rate, weight_decay, etc. If it is specific to SGD, it should check that the optimizer is indeed SGD.

**Review by subutai — CHANGES_REQUESTED** [review]

File: nupic/research/frameworks/sigopt/mixins/imagenet_params.py (hunk 2 above)
Comment (mvacaporale) [reply nupic/research/frameworks/sigopt/mixins/imagenet_params.py:130]: I have it overwrite it for full control. I expect it to be used only for experiments within our imagenet project. I may make this more explicit. Possibly, I can move this mixin to the projects folder to imply it is custom made for those experiments and shouldn't be generally applied.

File: nupic/research/frameworks/sigopt/mixins/imagenet_params.py (hunk 3 above)
Comment (mvacaporale) [reply nupic/research/frameworks/sigopt/mixins/imagenet_params.py:27]: Older configs here mean any those within the Imagent projects folder. I can make this more clear.

File: nupic/research/frameworks/sigopt/mixins/one_cycle_lr_params.py (hunk 6 above)
Comment (mvacaporale) [reply nupic/research/frameworks/sigopt/mixins/one_cycle_lr_params.py:68]: Good eye. This should be covered. In multi-task experiments, the param `config["epochs"]` gets defined in the base class `SigOptExperiment`. Since the function you reference calls the super's `update_config_with_suggestion`, then the "epochs" will be populated by the time `start_epochs = ...` is ran.

File: nupic/research/frameworks/sigopt/mixins/one_cycle_lr_params.py (hunk 7 above)
Comment (mvacaporale) [reply nupic/research/frameworks/sigopt/mixins/one_cycle_lr_params.py:61]: As mentioned, that's in `SigOptExperiment.update_config_with_suggestion`

File: nupic/research/frameworks/sigopt/mixins/sgd_params.py (hunk 8 above)
Comment (mvacaporale) [reply nupic/research/frameworks/sigopt/mixins/sgd_params.py:31]: I thought it would be helpful to be explicit about what each mixin is intended to do. If one wants to optimize Adam parameters, then it's fairly straightforward to introduce a new mixin. I'll add a check for the class type.

**Review by subutai — APPROVED** [review]
